pkg/whiteflag/white_flag.go

```diff
@@ -0,0 +1,247 @@
+	var trunkVisited, trunkConfirmed, branchVisited, branchConfirmed bool
+	var trunkConfirmedIndex, branchConfirmedIndex milestone.Index
+
+	if _, trunkVisited = visited[headTxTrunkHash]; !trunkVisited {
+		if cachedTrunkTx = tangle.GetCachedTransactionOrNil(headTxTrunkHash); cachedTrunkTx == nil {
+			return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxTrunkHash)
+		}
+		defer cachedTrunkTx.Release()
+		trunkConfirmed, trunkConfirmedIndex = cachedTrunkTx.GetMetadata().GetConfirmed()
+	}
+
+	if headTxTrunkHash == headTxBranchHash {
+		branchVisited = trunkVisited
+		branchConfirmedIndex = trunkConfirmedIndex
+		branchConfirmed = trunkConfirmed
+		cachedBranchTx = cachedTrunkTx
+		// no need to load branch tx
+	} else {
+		if _, branchVisited = visited[headTxBranchHash]; !branchVisited {
+			if cachedBranchTx = tangle.GetCachedTransactionOrNil(headTxBranchHash); cachedBranchTx == nil {
+				return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxBranchHash)
+			}
+			defer cachedBranchTx.Release()
+			branchConfirmed, branchConfirmedIndex = cachedBranchTx.GetMetadata().GetConfirmed()
+		}
+	}
+
+	// verify that head and trunk txs are indeed tails
+	if !trunkVisited && !cachedTrunkTx.GetTransaction().IsTail() {
+		return fmt.Errorf("%w: trunk tx %s of bundle head tx %s is not a tail", ErrMilestoneApprovedInvalidBundle, headTxTrunkHash, bundleHeadTx.GetHash())
+	}
+
+	if !branchVisited && !cachedBranchTx.GetTransaction().IsTail() {
+		return fmt.Errorf("%w: branch tx %s of bundle head tx %s is not a tail", ErrMilestoneApprovedInvalidBundle, headTxBranchHash, bundleHeadTx.GetHash())
+	}
+
+	// here we reached a tail of which its past cone was already visited, therefore we include its bundle
+	if (trunkVisited || tangle.SolidEntryPointsContain(headTxTrunkHash) || trunkConfirmedIndex != milestoneIndex) ||
+		(branchVisited || tangle.SolidEntryPointsContain(headTxBranchHash) || branchConfirmedIndex != milestoneIndex) {
+		// if the bundle is invalid or a value spam bundle, we don't incorporate it as part of the mutations
+		bundle := cachedBundle.GetBundle()
+
+		// on a bundle not mutating any address, we simply mark it as visited
+		if bundle.IsValueSpam() {
+			visited[currentTx.GetHash()] = struct{}{}
+			stack.Remove(ele)
+			return nil
+		}
+
+		var conflicting bool
+
+		// contains the updated mutations from this bundle against the
+		// current mutations of the milestone's confirming cone (or previous ledger state).
+		// we only apply it to the milestone's confirming cone mutations if
+		// the bundle doesn't create any conflict.
+		patchedState := make(map[trinary.Hash]int64)
+
+		for addr, change := range bundle.GetLedgerChanges() {
+
+			// load state from milestone cone mutation or previous milestone
+			balance, has := wfConf.NewAddressState[addr]
+			if !has {
+				balanceStateFromPreviousMilestone, _, err := tangle.GetBalanceForAddressWithoutLocking(addr)
```

Comment:
We should at least add a comment that we have to ReadLockLedger before calling `ProcessStack`